Use libvmi consistently in storage tests

brybacki opened the pull request:
Signed-off-by: Bartosz Rybacki <brybacki@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

Use `libvmi` where possible in storage tests. Many tests already use this new lib, so using it here makes codebase more consistent (simpler to maintain). This is one of required steps to remove big universal utils.go.

With this PR a new factory functions were added to create a VM with provided DV or PVC. Also new function was used in a few places as an example. In a followup PR I'll try to replace all tests.RandomVM usages.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

xpivarc commented on tests/storage/storage.go:
```diff
@@ -1031,9 +1036,14 @@ var _ = SIGDescribe("Storage", func() {
+						libvmi.WithPersistentVolumeClaim(dataVolume.Name),
+						libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+						libvmi.WithNetwork(virtv1.DefaultPodNetwork()))...)
+				//tests.NewRandomVMIWithPVC(dataVolume.Name)
```
Is this correct?

brybacki replied:
I don't know, unfortunately it does not work yet. I plan on returning to work on it tomorrow.

I wanted to replace `NewRandomVMIWithPVC` and found a VM needs interface and network, if you see anything else is missing feel free to tell me.

xpivarc replied:
I don't think you need to specify the interface. It's needed only if you want to not do functional change here. 
Do you fail 100%? It seems to me you were just unlucky.

I actually have similar work locally. I will try to push it.

brybacki replied:
Yes, 100% failure when trying on local build. 
Cannot log in.

brybacki replied:
Ok, found the problem, it was the order of disks, so far I have rebased and push changes for testing. So I can see if everything else works, and I'll try to clean up the PR.

xpivarc replied:
Can you elaborate?

brybacki replied:
Disk with cloudInit has to be last, after the disk with PVC/DV. I am not sure why. 

https://github.com/kubevirt/kubevirt/blob/892314aa31091629c127c81aa9e64001bffe8386/tests/libvmi/factory.go#L77

brybacki replied:
This is fixed. The order of disks/boot order

xpivarc commented:
/retest

brybacki commented:
/retest-required

brybacki commented:
/sig code-quality

brybacki commented:
/test pull-kubevirt-fossa

brybacki commented:
/retest

akrejcir commented on tests/libvmi/factory.go:
```diff
@@ -47,27 +46,49 @@ func NewFedora(opts ...Option) *kvirtv1.VirtualMachineInstance {
+}
+
+// NewCirrosOptions instantiates default options needed for a CirrOS based VMI configuration
+func NewCirrosOptions(opts ...Option) []Option {
```
I find it unusual, that this function takes a slice parameter and only adds it to the resulting slice. It could just return the `cirrosOpts` and the calling function would add them to other parameters.
Is it important that the cloudInit disk is last?

brybacki replied:
The main reason for such a function is that cloudInit had to be the last disk. So it takes all the parameters and applies everything in order.

I'll rebase and test it again to make sure this is the reason it failed. If cloudInit does not need to be last then I'll change the code so it simply returns options.

xpivarc replied:
We could use `bootOrder` if we for whatever reason don't want to rely on the order(of options)...

brybacki replied:
ok, confirmed, the hacky cloudInit has to be last.
So, I would need to add a bootOrder to the disk? I need to check how to do this correctly so other disks, any disks added by end user are not in conflict.

brybacki replied:
I am not sure it will be easier or cleaner, I would need to provide a consistent way to order all boot devices when creating cirros, because we want the cloudInit to be the last. Of course it will be more **correct** than ordering it in yaml. 

Anyway, I am experimenting with it right now.

brybacki replied:
From my experiments it looks like Boot order only works if applied to all devices. So to correctly create Cirros I would have to get the vmi spec, then order all disks, and then make sure the cloudInit is last. I can add this.

brybacki replied:
I have actually changed it so it returns options. And I have also changed the way we add boot disk. Now the last one wins (replaces previous one) so we can have defaults and it can be easily replaced by user provided options.

akrejcir commented on tests/libvmi/factory.go:
```diff
@@ -47,27 +46,49 @@ func NewFedora(opts ...Option) *kvirtv1.VirtualMachineInstance {
+}
+
+// NewCirrosOptions instantiates default options needed for a CirrOS based VMI configuration
+func NewAlpineOptions(opts ...Option) []Option {
```
The `opts` parameter is only passed through. Can you remove it, and leave it up to the caller to append any other paremeters?

When you rebase this PR on the latest `main`, there is a new function `NewAlpineWithTestTooling`, which could also use this function.

brybacki replied:
Thanks, good idea.

brybacki replied:
Well, actually I would like to keep it as a Decorator function. 
So the caller provides parameters, for example as `opts`, and then wraps/decorates it into Alpine flavor:
```
opts : = someUserProvidedOptionsAnddisks()
vmi := New(RandName(DefaultVmiName), NewAlpineOptions(opts...))

```

That way options provided by caller always take precedence (for example one can provide a different ContainerImage, first one provided wins). That way we can also do some checks/validations/add things that have to go last - as in Cirros.

brybacki replied:
I have rebased and done some cleanup, WDYT?

akrejcir reviewed (commented):
Looks good. I just have two small comments.

akrejcir commented:
/lgtm

brybacki commented:
I probably messed up something need to check failing tests

brybacki commented:
> I probably messed up something need to check failing tests

No, it is not me: All failed tests show `no lspci`

xpivarc commented:
> what is wrong here? the test 3135 that failed here on storage lanes is passing locally on 1.21 and 2.23!

I am not saying this is related but I wonder if we can have a look at this:
`
{"component":"virt-launcher","level":"error","msg":"No disk capacity","pos":"manager.go:661","timestamp":"2022-05-05T17:57:10.822412Z"}
`

I see this message quite often and the level suggests to me it is not good. I think it might be because the source is not pvc and therefore we don't see capacity even though we set the capacity in the `volumesource`.

As far as the failure... It is cause by this:
 ```
"devices": {
                        "disks": [
                            {
                                "name": "disk1",
                                "disk": {
                                    "bus": "virtio"
                                }
                            },
                            {
                                "name": "disk0",
                                "disk": {
                                    "bus": "virtio"
                                }
                            },
                            {
                                "name": "emptydisk1",
                                "disk": {
                                    "bus": "virtio"
                                },
                                "serial": "FB-fb_18030C10002032"
                            }
                        ],
```

where volumes are

```
"volumes": [
                    {
                        "name": "disk1",
                        "cloudInitNoCloud": {
                            "networkData": "ethernets:\n  eth0:\n    addresses:\n    - fd10:0:2::2/120\n    dhcp4: true\n    gateway6: fd10:0:2::1\n    match: {}\n    nameservers:\n      addresses:\n      - 10.96.0.10\n      search:\n      - default.svc.cluster.local\n      - svc.cluster.local\n      - cluster.local\nversion: 2\n"
                        }
                    },
                    {
                        "name": "disk0",
                        "containerDisk": {
                            "image": "registry:5000/kubevirt/alpine-with-test-tooling-container-disk:devel",
                            "imagePullPolicy": "IfNotPresent"
                        }
                    },
                    {
                        "name": "emptydisk1",
                        "emptyDisk": {
                            "capacity": "1Gi"
                        }
                    }
                ],
                "networks": [
                    {
                        "name": "default",
                        "pod": {}
                    }
                ]
            },
```

brybacki commented:
> As far as the failure... It is cause by this:

I can see 3 and 3 disks, it looks good to me. What is the problem?

xpivarc commented:
> > As far as the failure... It is cause by this:
> 
> I can see 3 volumes and 3 disks, it looks good to me. What is the problem?

@brybacki 
Have a look at the order.

brybacki commented:
This is from some kind of volume resize logic. I see it often too. I suppose the level is not adequate. Maybe it should not be logged/used when there is no need to resize.
 `{"component":"virt-launcher","level":"error","msg":"No disk capacity","pos":"manager.go:661","timestamp":"2022-05-05T17:57:10.822412Z"}`
 
 @maya-r from what I know you are en expert in the "volume resize" logic. Is this message an error?

brybacki commented:
> > > As far as the failure... It is cause by this:
> > 
> > 
> > I can see 3 volumes and 3 disks, it looks good to me. What is the problem?
> 
> @brybacki Have a look at the order.

So you think the order may be important here for Alpine? I only had such problems with the Cirros and the hacky "hello world".

xpivarc commented:
> > > > As far as the failure... It is cause by this:
> > > 
> > > 
> > > I can see 3 volumes and 3 disks, it looks good to me. What is the problem?
> > 
> > 
> > @brybacki Have a look at the order.
> 
> So you think the order may be important here for Alpine? I only had such problems with the Cirros and the hacky "hello world".

Yes, it is important because it doesn't have anything to do with the quest. It is a matter of firmware(bios/efi). I am not sure why bios doesn't try another device though.

brybacki commented:
> > > > > As far as the failure... It is cause by this:
> > > > 
> > > > 
> > > > I can see 3 volumes and 3 disks, it looks good to me. What is the problem?
> > > 
> > > 
> > > @brybacki Have a look at the order.
> > 
> > 
> > So you think the order may be important here for Alpine? I only had such problems with the Cirros and the hacky "hello world".
> 
> Yes, it is important because it doesn't have anything to do with the quest. It is a matter of firmware(bios/efi). I am not sure why bios doesn't try another device though.

Thanks, now I understand the problem better. My original problem was not because CloudInit was supposed to be last, but because the boot disk was supposed to be `first`. I think there is a need to add an explicit way to specify a `boot ` disk in our test factory.

brybacki commented:
Changed -  disk0 now always has the bootOrder = 1.

brybacki commented:
/test pull-kubevirt-build

brybacki commented:
It was too quick to celebrate victory ;) I see that I need to analyze few more tests

brybacki commented:
/retest

brybacki commented:
Now finally everything works! ;)

brybacki commented:
@xpivarc WDYT now? I had to add one last change.

xpivarc commented:
/cc @EdDev

EdDev commented on tests/libvmi/factory.go:
```diff
@@ -43,48 +43,55 @@ func NewFedora(opts ...Option) *kvirtv1.VirtualMachineInstance {
 		WithContainerImage(cd.ContainerDiskFor(cd.ContainerDiskFedoraTestTooling)),
 	}
-	opts = append(fedoraOptions, opts...)
+	opts = append(opts, fedoraOptions...)
```
I'm unsure why you swapped them, but I think this is problematic.
The inputted options should have the chance to override any defaults, but now it is not the case.

brybacki replied:
You are totally right, and the last commit has the right order. 
It looks like I need to cleanup/split the commits

EdDev reviewed (changes requested):
> Ad an option to create VMs with user provided DataVolumes and PVCs.

typo: Add

The first commit seems to do much more than it describes in the message.
I do not understand the reasoning behind the changes (e.g extracting the options function) and I think some are changing the existing intended behavior.

Please split the changes and reason for them one at a time. it will help the reviewers focus better and understand where it came from.

EdDev reviewed (changes requested):
> Tests: Add boot order to boot device
> 
> Sometimes cloud init data or empty disk is ordered before the actual
> boot disk. The test VM takes really long to start or does not start at
> all. The test framework marks the disk0 with bootOrder=1 to make sure it
>  is always used first and as a boot disk.

The original logic included defaults which could be overridden by the input.
I think this changed when you swapped the order, such that the defaults have been added **after** the input options.

What I would expect to see is that the order of definition is what defined the boot order, and I think it was like that until now. I would prefer to be dependent on that and not on this boot order explicit setting. If a specific test wants to check the boot order, it can do so, but I prefer the others to not have to have it included.

I hope that after you clarify the first commit (and hopefully split it), we could understand better what happened and if we can solve it in another way.

brybacki commented:
> What I would expect to see is that the order of definition is what defined the boot order, and I think it was like that until now. I would prefer to be dependent on that and not on this boot order explicit setting. If a specific test wants to check the boot order, it can do so, but I prefer the others to not have to have it included.

Yes I am about to cleanup commits, and describe it better. I think the previous change to hide default CloudInit for cirros (hello world) made this harder. I had many problems with my PR before I understood what is the problem.

Previously user provided options where first. Then the defaults were applied (disk was only added if it was not already present). With this solution there is a problem that sometimes a user adds a boot disk (which works ok it is added before any default cloudInit, or default boot - like for cirros). And sometimes a user can add a cloud Init (that would not work as it would be before the bootDisk), and I have seen some Alpine tests failing on this. 

So I changed the order, that now defaults are added first with default boot disk, default cloudInit and so on. But the user has an option to override disks (that is why I have changed apped to replace) and other options.  And I think this order - the defaults first, the user provided options last is better. It will work also for memory/rng etc, since this are a simple assignment operators. As sometimes cloudInit is first and I noticed cirros / alpine did not boot correctly, I had to add explicit boot order. 

But, after your comments, I can see that first I need to cleanup the Commits, and then we can continue with the review.

brybacki commented:
This looks like a usual flakiness in the test during update. Maybe that can be improved by using one retry. 
```
    Operation cannot be fulfilled on virtualmachines.kubevirt.io "testvmi-5pbjx": the object has been modified; please apply your changes to the latest version and try again
```
/test pull-kubevirt-e2e-k8s-1.22-sig-storage

brybacki commented:
/test pull-kubevirt-e2e-kind-1.22-sriov

brybacki commented:
@EdDev thank you for your comment from May 23, it forced me to rethink what I did, and after a cleanup (that was really easy) the code is smaller and better!

It looks like previously I have accumulated some "cruft" because I tried a few different ways to solve my first problems.

EdDev commented on tests/libvmi/storage.go:
```diff
@@ -28,8 +30,26 @@ import (
 		diskName := "disk0"
-		vmi.Spec.Domain.Devices.Disks = append(vmi.Spec.Domain.Devices.Disks, newDisk(diskName, v1.DiskBusVirtio))
-		vmi.Spec.Volumes = append(vmi.Spec.Volumes, newContainerVolume(diskName, name))
+		replaceDisk(vmi, newDisk(diskName, v1.DiskBusVirtio))
```
If the disk is already there, why would you like to replace it? I guess I am asking why `addDisk` is not a fit.

brybacki replied:
That way we have single and simple `libvmi.NewAlpine(` that uses containerImage, but in tests one can provide a PVC or DV with different content. The same for Cirros/Fedora. 

I suppose we could have a full blown framework with API to add/replace disks. But here I went for the simplest solution that works in current tests.

EdDev replied:
I am not sure that what we had so far is good, but let me try to see if I get this right:
Until now, we had two options: Some build options added disk (without checking that another with the same name existed already) and some build options ignored cases where the disks or volumes have been there already.

Now, the suggestion is that all build options that handle storage will either add or replace/override what was there before.
This may be fine, but it may also cause confusion.

I can share what originally guided me, and you take it from there and see if it makes sense and works with the new additions:
- If we need to add a volume, we add it to a disk that may either be there or not. If it is there already, we do nothing. If it is not there, we add it. More or less a lazy creation of the dependency.
- There are disks for container images and for cloud-init at the moment. (looks like the container image one is not following this rule, which I guess is bad).

So if we want to override the disk or volume, we wither need to choose which pattern makes more sense.
I guess another option is to explicitly require to override everything (e.g. `WithCleanDisks()`) or as you suggested, to always override (last one wins). But what happens when we have multiple volumes on the same disk?

I think you are more qualified to answer this one, so I will follow with what you think it fits better. We just need to make sure that nothing got broken by this.

brybacki replied:
All the options like Memory or Rng are written in a way that last one wins, so if we have a default, a user can always override it. This is also the reason why I think it is good to have the semantics of "last wins". In my opinion `WithXyz` should just set the value, and the last one should win. 

With ContainerImage before this PR it was strange because it just added a new disk with the same name. I am not sure if that is even valid. If we add only when the disk is not there then we just take the first value, which is valid but is not useful and not easy to use. I would need to add `NewAlpineWithoutDisk`. 

My only question is, what do you mean by " what happens when we have multiple volumes on the same disk"?

EdDev replied:
Not sure if I am not totally mistaken, the hierarchy in the VMI spec is super confusing (me).
I thought that a disk can have multiple volumes.

I think the difference from the other two examples you gave is that here we have things that depend on each other.
Adding a volume depends on a disk to exist.

Anyway, I am fine with whatever you prefer to have. If this will fail in some future scenario, we should be able to fix it.

brybacki replied:
I do not see any dependency here, 
I just add PVC, and I just add other alpine options, no dependencies at all.

EdDev commented on tests/storage/storage.go:
```diff
@@ -646,21 +649,21 @@ var _ = SIGDescribe("Storage", func() {
 			It("[test_id:3138]should start vmi multiple times", func() {
-				vmi = tests.NewRandomVMIWithPVC(tests.DiskAlpineHostPath)
-				tests.AddPVCDisk(vmi, "disk1", v1.DiskBusVirtio, tests.DiskCustomHostPath)
+				vmiSpec := libvmi.NewAlpine(libvmi.WithPersistentVolumeClaim(tests.DiskAlpineHostPath))
```
nit: Why did you changed the name of the variable `vmi`?

brybacki replied:
You are right, it is not consistent with the remaining tests.

For me it is much easier to differentiate the spec of vm that we define from the actual VM we get from k8s, by using different variables. And it is easier to spot a problem where someone expects to get actual vmi, but has 
`_, err := client.Get` in code, so is still using the original spec.

But it is not needed here. We can vote which way is better. Anyway i can remove it to make this PR contain only changes related to libvmi changes. The rest can be included in https://github.com/kubevirt/kubevirt/pull/7634.

EdDev replied:
For me it just raised questions. I will leave it to you and the maintainer to decide what is best.
Generally speaking, such changes cause the reviewer to look for the catch.

EdDev commented on tests/storage/storage.go:
```diff
@@ -646,21 +649,21 @@ var _ = SIGDescribe("Storage", func() {
 					if i == num {
 						By("Checking that the second disk is present")
-						Expect(console.LoginToAlpine(vmi)).To(Succeed())
+						Expect(console.LoginToAlpine(obj)).To(Succeed())
```
Seems like an unrelated fix (fits a different commit or PR). But I can look the other way :)

brybacki replied:
You are right, I should do it in https://github.com/kubevirt/kubevirt/pull/7634.

brybacki commented on tests/storage/storage.go:
```diff
@@ -247,6 +247,9 @@ var _ = SIGDescribe("Storage", func() {
 		})
 
 		Context("[rfe_id:3106][crit:medium][vendor:cnv-qe@redhat.com][level:component]with Alpine PVC", func() {
+			newRandomVMIWithPVC := func(claimName string) *virtv1.VirtualMachineInstance {
```
I have implemented it the same way as many other helpers in this storage.go file are. I think it is nice to have such small helpers defined close to usage. 

If you insist I can change it.

EdDev replied:
We experienced bugs when using closures when they have not been a must. They can bind to variables without you noticing.
But this is out of my domain, if the storage SIG prefers this way, I will not stand in your way.

EdDev reviewed (approved):
I think I got all the answers.
The discussion on the changes that may fit other changes, I leave for you to decide how to proceed.
Same for the semantics of the disks and volumes (replace vs keep as is). I do not think there is one perfect solution here, all have cons and pros.

Thanks for this cleanup!

brybacki commented:
@xpivarc wdyt? is it ready to go?

xpivarc commented on tests/libvmi/storage.go:
```diff
@@ -28,8 +30,26 @@ import (
+// WithPersistentVolumeClaim specifies the name of the PersistentVolumeClaim to be used.
+func WithPersistentVolumeClaim(name string) Option {
+	return func(vmi *kvirtv1.VirtualMachineInstance) {
+		diskName := "disk0"
```
I don't understand why we went with static names. It is not flexible and the usage of our options is very limited by this.

brybacki replied:
It was like that for container images, maybe the assumption was it will always be a single disk? No idea. 
My first idea was to make a small PR to add PVC and DV. Without any refactoring. So I did not change it.

It would require adding name to every call site `WithPersistentVolumeClaim(diskName, pvcName)`. But I think that will be easy.

xpivarc reviewed (commented):
I would like to see how can we shape the API in a way that test contributors will not make mistakes or the API will do what is expected.

brybacki commented on tests/storage/storage.go:
```diff
@@ -220,11 +220,11 @@ var _ = SIGDescribe("Storage", func() {
 				By("Creating VMI with faulty disk")
-				vmi := tests.NewRandomVMIWithPVC(pvc.Name)
-				_, err := virtClient.VirtualMachineInstance(util.NamespaceTestDefault).Create(vmi)
+				vmi := libvmi.NewCirros(libvmi.WithPersistentVolumeClaim(pvc.Name))
```
Not the same? what is missing>

brybacki commented:
> I would like to see how can we shape the API in a way that test contributors will not make mistakes or the API will do what is expected.

My first idea was that I am "just adding PVC and DV" in a small PR. But now I think it might be better to improve the API design. 

> I am just probably missing something but why do we want to replace? I would even go as far as panicking because it is a programmatical error.

We do replace the Memory, the Rng, all the other options. Some defaults are set, but the user can override them. That is how I see it. Why cannot we replace a disk? I see only one problem here, we always use one hardcoded disk name. So it is impossible to add more disks. And it requires a user to know the hardcoded name.

brybacki commented:
/test pull-kubevirt-code-lint

Is it a flaky unit test?
```
//pkg/inotify-informer:go_default_test: Inotify When watching files in a directory should detect multiple creations and deletions expand_less	
```

brybacki commented:
/test pull-kubevirt-unit-test

brybacki commented:
/test pull-kubevirt-e2e-k8s-1.23-operator

brybacki commented:
@EdDev @xpivarc can you review the newest set of changes?

EdDev commented on tests/libvmi/factory.go:
```diff
@@ -56,25 +56,27 @@ func NewCirros(opts ...Option) *kvirtv1.VirtualMachineInstance {
+	return New(alpineOpts...)
+}
+
+func AlpineOptions() []Option {
```
I do not understand why this is needed. It also looks odd, as these options have nothing to do with Alpine really.

brybacki replied:
this are default Alpine options that you can use when creating alpine VMs with some PVCs, DVs and other disks, this can also be reused in the factory functions

EdDev reviewed (commented):
First commit has some message leftovers:
```
# Conflicts:
#	tests/libvmi/storage.go
```

brybacki commented on tests/storage/storage.go:
```diff
@@ -248,6 +253,13 @@ var _ = SIGDescribe("Storage", func() {
+			newRandomVMIWithPVC := func(claimName string) *virtv1.VirtualMachineInstance {
+				disk := []libvmi.Option{libvmi.WithPersistentVolumeClaim("disk0", claimName)}
+
+				return libvmi.New(append(
+					disk,
+					libvmi.AlpineOptions()...)...)
```
NewAlpine has the default containerDisk, and many previous comments did not like the "override" behavior so I changed that. I changed so now one can add disk the same way as other options.

brybacki replied:
so you think it is better to repeat all the Alpine default options every time I create a new VM?

brybacki replied:
ok, done the simplest thing, inlined all the options from AlpineOptinos and it looks good.

EdDev commented:
Looks good. Can you remove the changes in the `factory.go` file? I do not think we need them now.

brybacki commented:
> Looks good. Can you remove the changes in the `factory.go` file? I do not think we need them now.

Good idea, done

brybacki commented:
@xpivarc can you take a look again, approve?

mhenriks commented:
/approve
